src/safeposix/syscalls/fs_calls.rs

```diff
@@ -1847,58 +1924,79 @@ impl Cage {
-                                //move position forward by the number of bytes we've written
-                                normalfile_filedesc_obj.position = position + byteswritten;
-                                newposition = normalfile_filedesc_obj.position;
-                                if newposition > normalfile_inode_obj.size {
-                                    normalfile_inode_obj.size = newposition;
-                                    drop(inodeobj);
-                                    drop(fileobject);
-                                    log_metadata(&FS_METADATA, normalfile_filedesc_obj.inode);
-                                } //update file size if necessary
-
-                                byteswritten as i32
-                            } else {
-                                0 //0 bytes written, but not an error value
-                                  // that can/should be passed to the user
+                            // Write `count` bytes from `buf` to the file at `position` using
+                            // `writeat` function, which returns the number of bytes written when
+                            // successful, else panics.
+                            let byteswritten = fileobject.writeat(buf, count, position).unwrap();
+                            // Move position forward by the number of bytes we've written
+                            normalfile_filedesc_obj.position = position + byteswritten;
+                            // Update the file size if necessary
+                            if normalfile_filedesc_obj.position > normalfile_inode_obj.size {
+                                normalfile_inode_obj.size = normalfile_filedesc_obj.position;
+                                drop(inodeobj);
+                                drop(fileobject);
+                                log_metadata(&FS_METADATA, normalfile_filedesc_obj.inode);
                             }
+                            // Return the number of bytes written
+                            byteswritten as i32
                         }
 
+                        // For `CharDev` type inode, the writing happens to the Character Device
+                        // file, with each device type returning the `count` number of bytes that
+                        // are to be written.
                         Inode::CharDev(ref char_inode_obj) => {
+                            // The `_write_chr_file` helper function typically does not write
+                            // anything to the device and simply returns the bytes count.
                             self._write_chr_file(&char_inode_obj, buf, count)
                         }
 
+                        // A Sanity check is added to make sure that there is no such case when the
+                        // fd type is "File" and the inode type is "Socket". This state is ideally
+                        // not possible, so we panic in such cases.
```

Comment:
since this case should never happen, should we check this at all? @rennergade

Reply:
I think we should add this case as well because we should showcase the results for all 4 types of inode objects (File, Socket, Dir, CharDev), and based on the conditions, return error/panic/process it.